### ISSUE TITLE
Show contacts from directory

### ISSUE THREAD

**Issue by CyberKiller40** [jdgeenen/gcontactsync#103]
The enterprise google apps allows to set up a users directory, which is shared between all users in the organization (e.g. stuff like email/phones to all employees, etc).

Would it be possible for the gContactSync plugin to read these addresses too?

**Comment by jdgeenen** [comment]
It looks very similar to the contacts API that gContactSync uses and it would probably work with very few changes.  However, if I'm reading [their documentation](https://developers.google.com/admin-sdk/domain-shared-contacts/#Feeds) correctly I think only domain administrators have access to the feed:

> The Shared Contacts Data API feeds is a private read/write feed that can be used by an adminstrator to view and manage domain's shared contacts. Since they are private, a programmer can access them only by using an authenticated request. That is, the request must contain an authentication token for a domain administrator. For more information about authenticating, see the developer's guide.

**Comment by CyberKiller40** [comment]
Yes, the directory contacts are read-only for normal users. Still, reading them is a very needed feature in an enterprise environment.

**Comment by aidalgol** [comment]
@CyberKiller40 The documentation page cited by @jdgeenen makes it sound like there is no read-only feed for normal users.  Either the documentation is unclear, or this is a gap in the API.  (A massive, crippling gap.)  Is there any (easy) way to test this?

**Comment by eesheesh** [comment]
The GAL feed is readable to all users in the domain unless the admin disabled it.

The URL is `https://www.google.com/m8/feeds/gal/domain.com/full`, where `domain.com` is the user's domain (can be obtained from the email address). It's the same as normal Contacts API resources, and paginated the same way too. It includes users, groups, calendar resources, and domain shared contacts.

There's some old example code that demonstrates that URL [here](https://github.com/google/gfw-deployments/blob/master/apps/shell/gal/gal_feed.sh), obviously it won't work because it uses ClientLogin which was shut down, but it can be updated to use OAuth.

As you've stated, there's no official documentation for this, but this should work and be easy to implement if you've already implemented Contacts API sync.

**Comment by sloncho** [comment]
I found a semi-workaround. Create a new group, then add all contacts from the GAL to that group. Yes, it's clumsy, as from time to time one have to repeat the above steps, when new people are added, but still it is better than nothing.

There is a small problem - somehow the names are not synced that way. Only the emails. Not good.

I guess reading the GAL feed is the way to go.

**Comment by eesheesh** [comment]
> There is a small problem - somehow the names are not synced that way. Only the emails. Not good.

This is because when you add a directory contact to My Contacts, it only adds the email address, and the rest of the information is added live by the Google Contacts web interface when you view it, so it's always fresh.

**Comment by jdgeenen** [comment]
The People API appears to support this: https://developers.google.com/people/api/rest/v1/people#sourcetype

**Comment by joelnoomi** [comment]
As a workaround, for now, I went to contacts.google.com logged in with my G Suite account, selected all contacts and then selected "Add contact(s)". After that, they were synced successfully by gContactSync. It would be great to have official support for shared contacts from the directory though.

**Comment by netdragonboberb** [comment]
The workaround of manual adding is clunky. A sync would still be superior.
